src/main/java/io/cdap/plugin/sap/transformer/ODataEntryToRecordTransformer.java

```diff
@@ -0,0 +1,419 @@
+    } catch (EdmSimpleTypeException e) {
+      throw new UnexpectedFormatException(String.format("Unsupported value for '%s' field: '%s'", fieldName, value), e);
+    }
+  }
+
+  private String extractDuration(String fieldName, BigDecimal decimal) {
+    try {
+      int precision = decimal.precision();
+      int scale = decimal.scale();
+      return EdmDuration.getInstance().valueToString(decimal, true, null, precision, scale, true);
+    } catch (EdmPrimitiveTypeException e) {
+      String errorMessage = String.format("Unsupported value for '%s' field: '%s'", fieldName, decimal);
+      throw new UnexpectedFormatException(errorMessage, e);
+    }
+  }
+
+  private int extractTimeMillis(Object value) {
+    long nanos = value instanceof GregorianCalendar
+      ? ((GregorianCalendar) value).toZonedDateTime().toLocalTime().toNanoOfDay()
+      : ((Timestamp) value).toLocalDateTime().toLocalTime().toNanoOfDay();
+    return Math.toIntExact(TimeUnit.NANOSECONDS.toMillis(nanos));
+  }
+
+  private long extractTimeMicros(Object value) {
+    long nanos = value instanceof GregorianCalendar
+      ? ((GregorianCalendar) value).toZonedDateTime().toLocalTime().toNanoOfDay()
+      : ((Timestamp) value).toLocalDateTime().toLocalTime().toNanoOfDay();
+    return TimeUnit.NANOSECONDS.toMicros(nanos);
+  }
+
+  private long extractTimestampMillis(Object value) {
+    Instant instant = value instanceof Calendar ? ((Calendar) value).toInstant() : ((Timestamp) value).toInstant();
+    long millis = TimeUnit.SECONDS.toMillis(instant.getEpochSecond());
+    return Math.addExact(millis, TimeUnit.NANOSECONDS.toMillis(instant.getNano()));
+  }
+
+  private long extractTimestampMicros(Object value) {
+    Instant instant = value instanceof Calendar ? ((Calendar) value).toInstant() : ((Timestamp) value).toInstant();
+    long micros = TimeUnit.SECONDS.toMicros(instant.getEpochSecond());
+    return Math.addExact(micros, TimeUnit.NANOSECONDS.toMicros(instant.getNano()));
+  }
+
+  private byte[] extractDecimal(String fieldName, Object value, Schema schema) {
+    int schemaPrecision = schema.getPrecision();
+    int schemaScale = schema.getScale();
+    BigDecimal decimal = extractBigDecimal(value, schema);
+    if (decimal.precision() > schemaPrecision) {
+      throw new UnexpectedFormatException(
+        String.format("Field '%s' has precision '%s' which is higher than schema precision '%s'.",
+                      fieldName, decimal.precision(), schemaPrecision));
+    }
+
+    if (decimal.scale() > schemaScale) {
+      throw new UnexpectedFormatException(
+        String.format("Field '%s' has scale '%s' which is not equal to schema scale '%s'.",
+                      fieldName, decimal.scale(), schemaScale));
+    }
+
+    return decimal.setScale(schemaScale).unscaledValue().toByteArray();
+  }
+
+  private BigDecimal extractBigDecimal(Object value, Schema schema) {
+    if (value instanceof BigDecimal) {
```

Comment:
this is confusing to me, when can the value be each of these different types?

Reply:
According to the [EdmPrimitiveType](https://olingo.apache.org/javadoc/odata4/org/apache/olingo/commons/api/edm/EdmPrimitiveType.html) `EDM.Decimal` can be represented by multiple Java types in Olingo V4: BigDecimal, BigInteger, Double, Float, Byte, Short, Integer, Long.


`EDM.Decimal` will be mapped to BigDecimal by Olingo if JSON response contains type control information [`@odata.type`](https://docs.oasis-open.org/odata/odata-json-format/v4.01/csprd05/odata-json-format-v4.01-csprd05.html#sec_ControlInformationtypeodatatype) for a decimal property. `application/json;odata.metadata=full` accept header allows to including control information to the response. However, not all SAP OData services support it.

I faced this issue testing the plugin with [SAP Gateway Demo System](https://blogs.sap.com/2017/12/05/new-sap-gateway-demo-system-available/).

Reply:
Corresponding Javadoc added.